packages/core/src/messagingProtocols/proteus/ProteusService/ProteusService.ts

```diff
@@ -173,4 +178,90 @@ export class ProteusService {
       state: response.errored ? PayloadBundleState.CANCELLED : PayloadBundleState.OUTGOING_SENT,
     };
   }
+
+  public async encrypt(
+    plainText: Uint8Array,
+    users: UserPreKeyBundleMap | UserClients,
+    domain: string = '',
+  ): Promise<{missing: UserClients; encrypted: OTRRecipients<Uint8Array>}> {
+    const {sessions, userClients} = await this.mapRecipantsToSessionsAndClients(users, domain, true);
+
+    const payload = await this.coreCryptoClient.proteusEncryptBatched(sessions, plainText);
+
+    return this.extractEncryptedAndMissingFromBatchedPayload(payload, userClients);
+  }
+
+  private async mapRecipantsToSessionsAndClients(
+    users: UserPreKeyBundleMap | UserClients,
+    domain: string = '',
+    createSession = false,
+  ) {
+    const userClients = filterUserClientsWithoutPreKey(users);
+    const sessions: string[] = [];
+
+    for (const userId in users) {
+      const clientIds = userClients[userId];
+
+      for (const clientId of clientIds) {
+        const sessionId = this.cryptographyService.constructSessionId(userId, clientId, domain);
+
+        if (createSession) {
+          const sessionExists = (await this.coreCryptoClient.proteusSessionExists(sessionId)) as unknown as boolean;
+
+          if (!sessionExists) {
+            const userQualifiedId = {id: userId, domain};
+            await this.createSession(userQualifiedId, clientId);
+          }
+        }
+
+        sessions.push(sessionId);
+      }
+    }
+
+    return {sessions, userClients};
+  }
+
+  private extractEncryptedAndMissingFromBatchedPayload(
+    payload: Map<string, Uint8Array>,
+    users: UserClients,
+  ): {missing: UserClients; encrypted: OTRRecipients<Uint8Array>} {
+    const encrypted: OTRRecipients<Uint8Array> = {};
+    const missing: UserClients = {};
+
+    const userClientsArr = Object.entries(users);
+    for (const [userId, clientIds] of userClientsArr) {
+      for (const clientId of clientIds) {
+        const sessionId = this.cryptographyService.constructSessionId(userId, clientId);
+
+        if (payload.has(sessionId)) {
+          encrypted[userId] ||= {};
+          encrypted[userId][clientId] = payload.get(sessionId)!;
+        } else {
+          missing[userId] ||= [];
+          missing[userId].push(clientId);
+        }
+      }
+    }
+
+    return {encrypted, missing};
+  }
```

Comment:
Not sure how do we want to keep track of missed (missing) clients with `proteusEncryptBatched` method... For now I've assumed that the clients that were included in recipients but are not included in returned batched payload are missing. 🤔

Reply:
I'm lacking the understanding of the behavior of `proteusEncryptBatched` when some sessions are missing. 
If it doesn't error and just returns the payload that it could encrypt, then this sounds like the perfect behavior to detect missings :)